File: michangsheng/ItemSystem/items/alchemy/tables.py
from typing import Dict, List, Iterator, Tuple, Optional
from items.constants import WuPingType
from items.loaders.items import items
from items.models.danyao import Herbs, YaoXing


class Tables:
    """草药药性速查表"""

    def __init__(self):
        self._initialed = False
        self.intro: Dict[YaoXing, List] = {}
        self.main: Dict[YaoXing, List] = {}
        self.deputy: Dict[YaoXing, List] = {}

    def _load(self):
        if self._initialed:
            return

        for item in items.filter_by_type(WuPingType.草药):
            herb = Herbs(item.id)
            for kind in ["intro", "main", "deputy"]:
                base: YaoXing = getattr(herb, kind)
                container: Dict[YaoXing, List] = getattr(self, kind)
                for num in range(1, 11):
                    container.setdefault(base * num, []).append((num, herb))

        self._initialed = True

    def list_possible_intro(
        self, key: YaoXing, fuzzy: bool = False, max_num: Optional[int] = None
    ) -> Iterator[Tuple[int, Herbs]]:
        """列举所有可以达到该需求的组合"""
        self._load()
        yield from filter_by_num(self.intro.get(key, []), max_num=max_num)

        if fuzzy:
            for possible in self.intro.keys():
                if key.kind == possible.kind and key.intensity < possible.intensity:
                    yield from filter_by_num(self.intro[possible], max_num=max_num)

    def list_possible_main(
        self, key: YaoXing, fuzzy: bool = False, max_num: Optional[int] = None
    ) -> Iterator[Tuple[int, Herbs]]:
        """列举所有可以达到该需求的组合"""
        self._load()
        yield from filter_by_num(self.main.get(key, []), max_num=max_num)

        if fuzzy:
            for possible in self.main.keys():
                if key.kind == possible.kind and key.intensity < possible.intensity:
                    yield from filter_by_num(self.main[possible], max_num=max_num)

    def list_possible_deputy(
        self, key: YaoXing, fuzzy: bool = False, max_num: Optional[int] = None
    ) -> Iterator[Tuple[int, Herbs]]:
        """列举所有可以达到该需求的组合"""
        self._load()
        yield from filter_by_num(self.deputy.get(key, []), max_num=max_num)

        if fuzzy:
            for possible in self.deputy.keys():
                if key.kind == possible.kind and key.intensity < possible.intensity:
                    yield from filter_by_num(self.deputy[possible], max_num=max_num)


def filter_by_num(
    iterator: Iterator[Tuple[int, Herbs]], max_num: Optional[int]
) -> Iterator[Tuple[int, Herbs]]:
    if max_num is None:
        yield from iterator
    else:
        for combination in iterator:
            if combination[0] <= max_num:
                yield combination


tables = Tables()

File: michangsheng/FriendlyLianDan/liandan/Tables.cs
using System.Collections.Generic;
using ItemSystem;
using ItemSystem.Models;

namespace FriendlyLianDan
{

    public class Combination
    {
        public int num;
        public Herbs herbs;

        public Combination(int num, Herbs herbs)
        {
            this.num = num;
            this.herbs = herbs;
        }

        public override string ToString()
        {
            return $"{num} * {herbs}";
        }
    }

    public class Tables
    {
        private static Tables instance = null;
        private static readonly object _lock = new object();
        private static int threshold = 11;

        public Dictionary<YaoXing, List<Combination>> intro;
        public Dictionary<YaoXing, List<Combination>> main;
        public Dictionary<YaoXing, List<Combination>> deputy;

        public static IEnumerable<Combination> listPossibleIntro(YaoXing key, bool fuzzy = false, int maxNum = -1)
        {
            if (Instance.intro.ContainsKey(key))
            {
                foreach (var item in filterByNum(Instance.intro[key], maxNum))
                {
                    yield return item;
                }
            }

            if (fuzzy)
            {
                foreach (var possibleKey in Instance.intro.Keys)
                {
                    if (key.kind == possibleKey.kind && key.intensity < possibleKey.intensity)
                    {
                        foreach (var item in filterByNum(Instance.intro[possibleKey], maxNum))
                        {
                            yield return item;
                        }
                    }
                }
            }
        }

        public static IEnumerable<Combination> listPossibleMain(YaoXing key, bool fuzzy = false, int maxNum = -1)
        {
            if (Instance.main.ContainsKey(key))
            {
                foreach (var item in filterByNum(Instance.main[key], maxNum))
                {
                    yield return item;
                }
            }

            if (fuzzy)
            {
                foreach (var possibleKey in Instance.main.Keys)
                {
                    if (key.kind == possibleKey.kind && key.intensity < possibleKey.intensity)
                    {
                        foreach (var item in filterByNum(Instance.main[possibleKey], maxNum))
                        {
                            yield return item;
                        }
                    }
                }
            }
        }

        public static IEnumerable<Combination> listPossibleDeputy(YaoXing key, bool fuzzy = false, int maxNum = -1)
        {
            if (Instance.deputy.ContainsKey(key))
            {
                foreach (var item in filterByNum(Instance.deputy[key], maxNum))
                {
                    yield return item;
                }
            }

            if (fuzzy)
            {
                foreach (var possibleKey in Instance.deputy.Keys)
                {
                    if (key.kind == possibleKey.kind && key.intensity < possibleKey.intensity)
                    {
                        foreach (var item in filterByNum(Instance.deputy[possibleKey], maxNum))
                        {
                            yield return item;
                        }
                    }
                }
            }
        }

        private static IEnumerable<Combination> filterByNum(IEnumerable<Combination> iterable, int maxNum = -1)
        {
            foreach (var item in iterable)
            {
                if (maxNum == -1 || item.num <= maxNum)
                {
                    yield return item;
                }
            }
        }

        private Tables()
        {
            string[] fields = new string[] { "intro", "main", "deputy" };

            intro = new Dictionary<YaoXing, List<Combination>>();
            main = new Dictionary<YaoXing, List<Combination>>();
            deputy = new Dictionary<YaoXing, List<Combination>>();

            foreach (var item in ItemSystem.Shims.Items.FilterByType(WuPingType.草药))
            {
                Herbs herbs = new Herbs(item.id);        
                foreach (string field in fields)
                {
                    var herbsYaoXing = (YaoXing)typeof(Herbs).GetField(field).GetValue(herbs);
                    var container = (Dictionary<YaoXing, List<Combination>>)typeof(Tables).GetField(field).GetValue(this);
                    for (int i = 1; i <= threshold; i++)
                    {
                        var k = herbsYaoXing * i;
                        if (!container.ContainsKey(k))
                        {
                            container.Add(k, new List<Combination>());
                        }
                        container[k].Add(new Combination(i, herbs));
                    }
                }
            }
        }

        public static Tables Instance
        {
            get
            {
                if (instance == null)
                {
                    lock (_lock)
                    {
                        if (instance == null)
                        {
                            instance = new Tables();
                        }
                    }
                }
                return instance;
            }
        }
    }
}